query-engine/prisma-models/src/builders/internal_dm_builder.rs

```diff
@@ -373,21 +373,26 @@ fn relation_placeholders(datamodel: &dml::Datamodel) -> Vec<RelationPlaceholder>
                 },
             };
 
-            result.push(RelationPlaceholder {
+            let placeholder = RelationPlaceholder {
                 name: name.clone(),
                 model_a,
                 model_b,
                 field_a,
                 field_b,
                 manifestation,
-            })
+            };
+
+            // Skip duplicate placeholders
+            if !result.contains(&placeholder) {
+                result.push(placeholder);
+            }
```

Comment:
Using a `HashSet` or a `HashMap` was creating a waterfall of `Eq` and `Hash` requirements for lots of structs.
I decided to use `Vec::contains` so that we could reuse the existing `PartialEq` implementations. Let me know if that sucks too much.